monai/transforms/io/array.py

```diff
@@ -241,14 +245,46 @@ def __call__(self, filename: Union[Sequence[PathLike], PathLike], reader: Option
             raise ValueError("`meta_data` must be a dict.")
         # make sure all elements in metadata are little endian
         meta_data = switch_endianness(meta_data, "<")
-        if self.ensure_channel_first:
-            img_array = EnsureChannelFirst()(img_array, meta_data)
 
-        if self.image_only:
-            return img_array
         meta_data[Key.FILENAME_OR_OBJ] = f"{ensure_tuple(filename)[0]}"  # Path obj should be strings for data loader
-
-        return img_array, meta_data
+        img = self.join_im_and_meta(img_array, meta_data)
+        if self.ensure_channel_first:
+            img = EnsureChannelFirst()(img)
+        return img
+
+    @staticmethod
+    def join_im_and_meta(im, meta: dict):
+        img = torch.as_tensor(im)
+
+        # if not tracking metadata, return torch.Tensor
+        if not get_track_meta() or meta is None:
+            return img
+
+        if "affine" in meta:
+            meta["affine"] = torch.as_tensor(meta["affine"])
+
+        # TODO: delete extra metadata
+        for i in range(8):
+            for k in ("dim", "pixdim"):
+                if f"{k}[{i}]" in meta:
+                    del meta[f"{k}[{i}]"]
+        for k in (
```

Comment:
Maybe move these keys to a global list somewhere (`TRANSFORM_KEYS`)? What we're doing is removing the transform values because they don't make sense here, this might be done elsewhere and it would be good to know what keys these are and in case add more for other formats since these are for Nifti. Then you can have:

```
remove_keys(meta, *TRANSFORM_KEYS)
```